File: samples/python/starops_sdk_samples/examples/chat_from_file.py
#!/usr/bin/env python3
"""
从文件加载请求示例
Load requests from file example

Usage:
    python -m starops_sdk_samples.examples.chat_from_file -file ../../requests/starops/entity.json
    python -m starops_sdk_samples.examples.chat_from_file -dir ../../requests/starops/
"""

import argparse
import asyncio
import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path
from typing import Optional

from ..client import AgentClient, Config, SDKException, SimplePrinter, EventPrinter, InteractiveHandler, InteractiveResponse


def parse_args():
    parser = argparse.ArgumentParser(description="从文件加载请求")
    parser.add_argument("-file", dest="file_path", help="请求 JSON 文件路径")
    parser.add_argument("-dir", dest="dir_path", help="请求文件目录")
    parser.add_argument("-simple", action="store_true", help="简洁模式")
    parser.add_argument("-output", default="../../requests/output", help="输出目录")
    parser.add_argument("-simulate-error", dest="simulate_error", action="store_true", help="模拟网络断连，测试重试逻辑")
    return parser.parse_args()


async def process_file(client: AgentClient, file_path: str, output_dir: str, simple_mode: bool):
    try:
        # Load request file
        with open(file_path, "r", encoding="utf-8") as f:
            req_file = json.load(f)

        # Extract message
        message = extract_message(req_file)
        if not message:
            print("⚠️ 文件中没有消息内容")
            return

        file_name = Path(file_path).name
        print(f"📄 文件: {file_name}")
        print(f"💬 消息: {truncate(message, 60)}")

        # Create thread
        thread_id = client.create_thread()

        # Create output file
        output_file = create_output_file(file_path, output_dir)

        # Write request info
        write_output(output_file, f"# Request: {file_name}")
        write_output(output_file, f"# Time: {datetime.now().isoformat()}")
        write_output(output_file, f"# ThreadID: {thread_id}")
        write_output(output_file, f"# Message: {message}\n")

        print("-" * 60)

        # Extract variables
        variables = req_file.get("variables", {})

        # Send request
        start_time = time.time()
        simple_printer = SimplePrinter() if simple_mode else None
        event_printer = None if simple_mode else EventPrinter(print_raw_body=False, print_parsed=True)
        interactive_handler = InteractiveHandler(client)
        event_index = 0

        events = client.chat_with_variables(thread_id, message, variables)
        while events is not None:
            try:
                event = await events.__anext__()
            except StopAsyncIteration:
                break

            event_index += 1

            if event.has_error():
                print(f"❌ 错误: {event.error}")
                write_output(output_file, f"[ERROR] {event.error}")
                break

            # Write raw event
            if event.raw_json:
                write_output(output_file, f"[EVENT {event_index}]\n{event.raw_json}\n")

            # 正常输出（先输出）
            if simple_mode:
                text = simple_printer.process_event(event)
                if text:
                    print(text, end="", flush=True)
            else:
                event_printer.print_event(event, event_index)

            # 检测交互事件（在输出之后）
            interactive_resp = _extract_interactive_event(event, interactive_handler)
            if interactive_resp:
                print("\n🔄 检测到交互事件，用户已响应...")
                events = interactive_handler.resume_chat(thread_id, interactive_resp, variables)
                event_index = 0

        elapsed = time.time() - start_time
        print()

        # Write final result
        if simple_mode and simple_printer:
            final_text = simple_printer.get_final_text()
            write_output(output_file, f"\n# Final Result:\n{final_text}")
            print(f"📄 最终文本:\n{final_text}")

        write_output(output_file, f"\n# Duration: {elapsed:.2f}s")
        print(f"⏱️  耗时: {elapsed:.2f}s")

        if output_file:
            output_file.close()
            print(f"📁 输出: {output_file.name}")

    except Exception as e:
        print(f"❌ 处理文件失败: {e}")


async def process_directory(client: AgentClient, dir_path: str, output_dir: str, simple_mode: bool):
    files = list(Path(dir_path).glob("*.json"))
    if not files:
        print(f"⚠️ 目录中没有 JSON 文件: {dir_path}")
        return

    print(f"📁 找到 {len(files)} 个请求文件\n")

    for i, file_path in enumerate(files):
        print(f"━━━ [{i + 1}/{len(files)}] {file_path.name} ━━━")
        await process_file(client, str(file_path), output_dir, simple_mode)
        print()

    print(f"✅ 处理完成，共 {len(files)} 个文件")


def extract_message(req_file: dict) -> str:
    messages = req_file.get("messages", [])
    if messages and messages[0].get("contents"):
        contents = messages[0]["contents"]
        if contents and contents[0].get("value"):
            return contents[0]["value"]
    return ""


def create_output_file(input_file: str, output_dir: str):
    try:
        os.makedirs(output_dir, exist_ok=True)
        base_name = Path(input_file).stem
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        output_path = os.path.join(output_dir, f"{base_name}-{timestamp}.log")
        return open(output_path, "w", encoding="utf-8")
    except Exception:
        return None


def write_output(f, content: str):
    if f:
        f.write(content + "\n")


def truncate(s: str, max_len: int) -> str:
    if len(s) <= max_len:
        return s
    return s[:max_len] + "..."


async def main_async():
    args = parse_args()

    if not args.file_path and not args.dir_path:
        print("用法:")
        print("  -file <path>   处理单个文件")
        print("  -dir <path>    处理目录下所有 JSON 文件")
        sys.exit(1)

    print("🚀 VibeOps Chat - 从文件加载请求 (Python)")
    print("=" * 60)

    try:
        # Load configuration
        cfg = Config.load_from_env()

        if args.simulate_error:
            cfg.simulate_network_error = True
            print("⚠️  已启用网络断连模拟，将在收到首个事件后触发重试")

        # Create client
        client = AgentClient(cfg)

        # Ensure output directory exists
        os.makedirs(args.output, exist_ok=True)

        # Process files
        if args.dir_path:
            await process_directory(client, args.dir_path, args.output, args.simple)
        else:
            await process_file(client, args.file_path, args.output, args.simple)

    except SDKException as e:
        print(f"❌ 配置加载失败: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ 错误: {e}")
        sys.exit(1)


def _extract_interactive_event(event, handler: InteractiveHandler) -> Optional[InteractiveResponse]:
    """从 ChatEvent 中检测交互事件并处理用户响应"""
    if not event.raw_json:
        return None
    try:
        body = json.loads(event.raw_json)
        for msg in body.get("messages", []):
            for evt in msg.get("events", []):
                if InteractiveHandler.is_interactive_event(evt):
                    call_id = msg.get("callId", "")
                    return handler.handle_event(evt, call_id)
    except Exception as e:
        print(f"⚠️ 交互事件解析失败: {e}")
    return None


def main():
    asyncio.run(main_async())


if __name__ == "__main__":
    main()

File: samples/typescript/src/examples/chat-from-file.ts
#!/usr/bin/env tsx
/**
 * 从文件加载请求示例
 * Load requests from file example
 *
 * Usage:
 *   npm run chat-from-file -- -file ../../sample-requests/entity.json
 *   npm run chat-from-file -- -dir ../../sample-requests/
 */

import * as fs from 'fs';
import * as path from 'path';
import { loadConfigFromEnv, AgentClient, SimplePrinter, EventPrinter, InteractiveHandler, InteractiveResponse, SDKException } from '../client/index.js';

interface Args {
  filePath?: string;
  dirPath?: string;
  simpleMode: boolean;
  outputDir: string;
  simulateError: boolean;
}

function parseArgs(): Args {
  const args: Args = {
    simpleMode: false,
    outputDir: '../logs',
    simulateError: false,
  };

  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    switch (argv[i]) {
      case '-file':
        args.filePath = argv[++i];
        break;
      case '-dir':
        args.dirPath = argv[++i];
        break;
      case '-simple':
        args.simpleMode = true;
        break;
      case '-output':
        args.outputDir = argv[++i];
        break;
      case '-simulate-error':
        args.simulateError = true;
        break;
    }
  }

  return args;
}

async function processFile(
  client: AgentClient,
  filePath: string,
  outputDir: string,
  simpleMode: boolean
): Promise<void> {
  try {
    // Load request file
    const content = fs.readFileSync(filePath, 'utf-8');
    const reqFile = JSON.parse(content);

    // Extract message
    const message = extractMessage(reqFile);
    if (!message) {
      console.log('⚠️ 文件中没有消息内容');
      return;
    }

    const fileName = path.basename(filePath);
    console.log(`📄 文件: ${fileName}`);
    console.log(`💬 消息: ${truncate(message, 60)}`);

    // Create thread
    const threadId = await client.createThread();

    // Create output file
    const outputFile = createOutputFile(filePath, outputDir);

    // Write request info
    writeOutput(outputFile, `# Request: ${fileName}`);
    writeOutput(outputFile, `# Time: ${new Date().toISOString()}`);
    writeOutput(outputFile, `# ThreadID: ${threadId}`);
    writeOutput(outputFile, `# Message: ${message}\n`);

    console.log('-'.repeat(60));

    // Extract variables
    const variables = reqFile.variables || {};

    // Send request
    const startTime = Date.now();
    const simplePrinter = simpleMode ? new SimplePrinter() : null;
    const eventPrinter = simpleMode ? null : new EventPrinter(false, true);
    const interactiveHandler = new InteractiveHandler(client);
    let eventIndex = 0;

    let events = client.chatWithVariables(threadId, message, variables);
    while (events) {
      let shouldContinue = false;
      for await (const event of events) {
        eventIndex++;

        if (event.error) {
          const err = event.error as SDKException;
          console.log(`❌ 错误: ${err.message}`);
          if (err.cause) {
            console.log(`   原因: ${(err.cause as Error).message || err.cause}`);
          }
          if ((err as any).context) {
            console.log(`   详情: ${JSON.stringify((err as any).context)}`);
          }
          writeOutput(outputFile, `[ERROR] ${err.message}`);
          shouldContinue = false;
          break;
        }

        // Write raw event
        if (event.rawJson) {
          writeOutput(outputFile, `[EVENT ${eventIndex}]\n${event.rawJson}\n`);
        }

        // 正常输出（先输出）
        if (simpleMode) {
          const text = simplePrinter!.processEvent(event);
          if (text) {
            process.stdout.write(text);
          }
        } else {
          eventPrinter!.printEvent(event, eventIndex);
        }

        // 检测交互事件（在输出之后）
        const interactiveResp = await extractInteractiveEvent(event, interactiveHandler);
        if (interactiveResp) {
          console.log('\n🔄 检测到交互事件，用户已响应...');
          events = interactiveHandler.resumeChat(threadId, interactiveResp, variables);
          eventIndex = 0;
          shouldContinue = true;
          break;
        }

        if (event.isDone) {
          shouldContinue = false;
          break;
        }
      }
      if (!shouldContinue) break;
    }

    const elapsed = Date.now() - startTime;
    console.log();

    // Write final result
    if (simpleMode && simplePrinter) {
      const finalText = simplePrinter.getFinalText();
      writeOutput(outputFile, `\n# Final Result:\n${finalText}`);
      console.log(`📄 最终文本:\n${finalText}`);
    }

    writeOutput(outputFile, `\n# Duration: ${elapsed}ms`);
    console.log(`⏱️  耗时: ${elapsed}ms`);

    if (outputFile) {
      outputFile.close();
    }
  } catch (e) {
    console.log(`❌ 处理文件失败: ${(e as Error).message}`);
  }
}

async function processDirectory(
  client: AgentClient,
  dirPath: string,
  outputDir: string,
  simpleMode: boolean
): Promise<void> {
  const files = fs.readdirSync(dirPath).filter((f) => f.endsWith('.json'));
  if (files.length === 0) {
    console.log(`⚠️ 目录中没有 JSON 文件: ${dirPath}`);
    return;
  }

  console.log(`📁 找到 ${files.length} 个请求文件\n`);

  for (let i = 0; i < files.length; i++) {
    console.log(`━━━ [${i + 1}/${files.length}] ${files[i]} ━━━`);
    await processFile(client, path.join(dirPath, files[i]), outputDir, simpleMode);
    console.log();
  }

  console.log(`✅ 处理完成，共 ${files.length} 个文件`);
}

function extractMessage(reqFile: Record<string, unknown>): string | undefined {
  const messages = reqFile.messages as Array<Record<string, unknown>> | undefined;
  if (messages && messages[0]) {
    const contents = messages[0].contents as Array<Record<string, unknown>> | undefined;
    if (contents && contents[0]?.value) {
      return contents[0].value as string;
    }
  }
  return undefined;
}

function createOutputFile(inputFile: string, outputDir: string): fs.WriteStream | null {
  try {
    fs.mkdirSync(outputDir, { recursive: true });
    const baseName = path.basename(inputFile, '.json');
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
    const outputPath = path.join(outputDir, `${baseName}-${timestamp}.log`);
    return fs.createWriteStream(outputPath);
  } catch {
    return null;
  }
}

function writeOutput(file: fs.WriteStream | null, content: string): void {
  if (file) {
    file.write(content + '\n');
  }
}

function truncate(s: string, maxLen: number): string {
  if (s.length <= maxLen) return s;
  return s.slice(0, maxLen) + '...';
}

async function main() {
  const args = parseArgs();

  if (!args.filePath && !args.dirPath) {
    console.log('用法:');
    console.log('  -file <path>   处理单个文件');
    console.log('  -dir <path>    处理目录下所有 JSON 文件');
    process.exit(1);
  }

  console.log('🚀 VibeOps Chat - 从文件加载请求 (TypeScript)');
  console.log('='.repeat(60));

  try {
    // Load configuration
    const cfg = await loadConfigFromEnv();

    if (args.simulateError) {
      cfg.simulateNetworkError = true;
      console.log('⚠️  已启用网络断连模拟，将在收到首个事件后触发重试');
    }

    // Create client
    const client = new AgentClient(cfg);

    // Ensure output directory exists
    fs.mkdirSync(args.outputDir, { recursive: true });

    // Process files
    if (args.dirPath) {
      await processDirectory(client, args.dirPath, args.outputDir, args.simpleMode);
    } else if (args.filePath) {
      await processFile(client, args.filePath, args.outputDir, args.simpleMode);
    }
  } catch (e) {
    if (e instanceof SDKException) {
      console.log(`❌ 配置加载失败: ${e}`);
    } else {
      console.log(`❌ 错误: ${(e as Error).message}`);
    }
    process.exit(1);
  }
}

async function extractInteractiveEvent(
  event: { rawJson?: string },
  handler: InteractiveHandler
): Promise<InteractiveResponse | null> {
  if (!event.rawJson) return null;
  try {
    const body = JSON.parse(event.rawJson);
    for (const msg of body.messages || []) {
      for (const evt of msg.events || []) {
        if (InteractiveHandler.isInteractiveEvent(evt)) {
          const callId = msg.callId || '';
          return await handler.handleEvent(evt, callId);
        }
      }
    }
  } catch (e) {
    console.log(`⚠️ 交互事件解析失败: ${(e as Error).message}`);
  }
  return null;
}

main();
